### PR TITLE
Allow Noam lr scheduler to run for more than max_steps

### PR DESCRIPTION
**Pull request by alancucki**
Signed-off-by: Adrian Lancucki <alancucki@users.noreply.github.com>

# What does this PR do ?

Allows to train for more than `max_steps` without lowering lr to `min_lr`.

Noam lr scheduler should decay lr indefinitely. The current implementation puts a hard limit at `max_steps` - the number of intended training epochs - after which it returns `min_lr`. This is contrary to the definition in the paper. Upon resuming from a ckpt `max_steps` are loaded from the ckpt, and if the user wishes to train for more epochs, they will unexpectedly get `min_lr`.

**Collection**: [Note which collection this PR will affect]

N/A

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by titu1994 — COMMENTED** [review]
This is fine, but note that if you're restoring a checkpoint, then it's parameters for max steps remains same as original, and the formula will be incorrect. I'd urge you to test this out in a unittests before merging

**Comment by alancucki** [comment]
Thanks for the quick comment. Added unittest which caught an edge case `warmup_steps == 0`.

After loading from a checkpoint, `max_steps` will remain same as original, but they will be ignored. I believe that the formula for calculating lr remains valid.

**Review by titu1994 — APPROVED** [review]
Fantastic, thanks !

**Review by titu1994 — DISMISSED** [review]

File: tests/core/test_optimizers_schedulers.py [hunk 1]
```diff
@@ -616,6 +617,48 @@ def test_CosineAnnealing(self):
...
+                assert policy1.get_last_lr()[0] <= self.INITIAL_LR
+            else:
+                assert self.MIN_LR < policy1.get_last_lr()[0] < self.INITIAL_LR
+                assert policy1.get_last_lr()[0] == policy2.get_last_lr()[0]
+
+            opt1.step()
+            opt2.step()
+            policy1.step()
+            policy2.step()
+
```
Comment (XuesongYang) [inline tests/core/test_optimizers_schedulers.py:661]: nit: it would be easier to maintain unit tests if decoupling warmup and no_warmup as two separate functions, for example, 

```
def test_NoamAnnealing_without_warmup(self):
    pass

def test_NoamAnnealing_with_warmup(self):
    pass
```

**Review by XuesongYang — CHANGES_REQUESTED** [review]

**Review by titu1994 — APPROVED** [review]
